File: client/app/src/commonMain/kotlin/com/jahirtrap/cconnect/data/remote/ClaudeApi.kt
package com.jahirtrap.cconnect.data.remote

import kotlinx.serialization.json.JsonElement

import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.floatOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

object ClaudeApi {

    data class Plugin(
        val name: String,
        val marketplace: String?,
        val version: String?,
        val scope: String?,
        val enabled: Boolean,
        val description: String? = null,
    )

    data class Marketplace(val name: String, val repo: String?)

    data class Skill(
        val name: String,
        val id: String,
        val description: String?,
        val plugin: String?,
        val pluginName: String?,
        val enabled: Boolean,
    )

    data class McpServer(val name: String, val type: String?, val detail: String?, val enabled: Boolean = true)

    data class Extensions(val plugins: List<Plugin>, val marketplaces: List<Marketplace>)

    data class CatalogPlugin(
        val name: String,
        val description: String?,
        val version: String?,
        val installed: Boolean,
    )

    data class Memory(val scope: String, val name: String, val description: String?)

    data class Memories(val global: List<Memory>, val project: List<Memory>)

    data class ActionResult(val ok: Boolean, val message: String)

    data class UsageWindow(val id: String, val percent: Float, val resetsAt: String?)

    data class Usage(val plan: String?, val windows: List<UsageWindow>, val error: String?)

    suspend fun usage(account: String? = null): Usage? {
        val query = account?.takeIf { it.isNotBlank() }?.let { mapOf("account" to it) } ?: emptyMap()
        val o = Http.get("/claude/usage", query)?.jsonObject ?: return null
        return Usage(
            plan = o["plan"]?.jsonPrimitive?.contentOrNull,
            windows = o["windows"]?.jsonArray?.map { el ->
                val w = el.jsonObject
                UsageWindow(
                    id = w["id"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    percent = w["percent"]?.jsonPrimitive?.floatOrNull ?: 0f,
                    resetsAt = w["resets_at"]?.jsonPrimitive?.contentOrNull,
                )
            }.orEmpty(),
            error = o["error"]?.jsonPrimitive?.contentOrNull,
        )
    }

    data class ServiceComponent(val name: String, val status: String)

    data class ServiceIncident(
        val name: String,
        val impact: String,
        val status: String,
        val latest: String?,
        val updatedAt: String?,
        val shortlink: String?,
    )

    data class ServiceStatus(
        val indicator: String,
        val description: String,
        val components: List<ServiceComponent>,
        val incidents: List<ServiceIncident>,
        val error: String?,
    )

    suspend fun status(): ServiceStatus? {
        val o = Http.get("/claude/status")?.jsonObject ?: return null
        return ServiceStatus(
            indicator = o["indicator"]?.jsonPrimitive?.contentOrNull ?: "none",
            description = o["description"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            components = o["components"]?.jsonArray?.mapNotNull { el ->
                val c = el.jsonObject
                ServiceComponent(
                    name = c["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
                    status = c["status"]?.jsonPrimitive?.contentOrNull ?: "operational",
                )
            }.orEmpty(),
            incidents = o["incidents"]?.jsonArray?.mapNotNull { el ->
                val i = el.jsonObject
                ServiceIncident(
                    name = i["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
                    impact = i["impact"]?.jsonPrimitive?.contentOrNull ?: "none",
                    status = i["status"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    latest = i["latest"]?.jsonPrimitive?.contentOrNull,
                    updatedAt = i["updated_at"]?.jsonPrimitive?.contentOrNull,
                    shortlink = i["shortlink"]?.jsonPrimitive?.contentOrNull,
                )
            }.orEmpty(),
            error = o["error"]?.jsonPrimitive?.contentOrNull,
        )
    }

    suspend fun userPrompt(): String? =
        Http.get("/claude/prompt")?.jsonObject?.get("text")?.jsonPrimitive?.contentOrNull

    suspend fun setUserPrompt(text: String): Boolean =
        Http.put("/claude/prompt", buildJsonObject { put("text", text) }) != null

    suspend fun projectPrompt(project: String): String? =
        Http.get("/claude/project-prompt", mapOf("project" to project))?.jsonObject?.get("text")?.jsonPrimitive?.contentOrNull

    suspend fun setProjectPrompt(project: String, text: String): Boolean =
        Http.put("/claude/project-prompt", buildJsonObject { put("project", project); put("text", text) }) != null

    suspend fun extensions(): Extensions? {
        val data = Http.get("/claude/plugins")?.jsonObject ?: return null
        val plugins = data["plugins"]?.jsonArray?.mapNotNull { el ->
            val o = el.jsonObject
            Plugin(
                name = o["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
                marketplace = o["marketplace"]?.jsonPrimitive?.contentOrNull,
                version = o["version"]?.jsonPrimitive?.contentOrNull,
                scope = o["scope"]?.jsonPrimitive?.contentOrNull,
                enabled = o["enabled"]?.jsonPrimitive?.booleanOrNull ?: true,
                description = o["description"]?.jsonPrimitive?.contentOrNull,
            )
        } ?: emptyList()
        val marketplaces = data["marketplaces"]?.jsonArray?.mapNotNull { el ->
            val o = el.jsonObject
            Marketplace(
                name = o["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
                repo = o["repo"]?.jsonPrimitive?.contentOrNull,
            )
        } ?: emptyList()
        return Extensions(plugins, marketplaces)
    }

    suspend fun skills(): List<Skill>? = Http.get("/claude/skills")?.jsonArray?.mapNotNull { el ->
        val o = el.jsonObject
        Skill(
            name = o["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
            id = o["id"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
            description = o["description"]?.jsonPrimitive?.contentOrNull,
            plugin = o["plugin"]?.jsonPrimitive?.contentOrNull,
            pluginName = o["plugin_name"]?.jsonPrimitive?.contentOrNull,
            enabled = o["enabled"]?.jsonPrimitive?.booleanOrNull ?: true,
        )
    }

    suspend fun skillFiles(plugin: String?, skillId: String): List<String> {
        val query = mutableMapOf("skill" to skillId)
        if (plugin != null) query["plugin"] = plugin
        return Http.get("/claude/skills/files", query)?.jsonObject?.get("files")?.jsonArray
            ?.mapNotNull { it.jsonPrimitive.contentOrNull } ?: emptyList()
    }

    fun skillFileUrl(plugin: String?, skillId: String, file: String = "SKILL.md"): String =
        "${Backend.baseUrl}/claude/skills/file?skill=${UrlCodec.encode(skillId)}" +
            (plugin?.let { "&plugin=${UrlCodec.encode(it)}" } ?: "") +
            "&file=${UrlCodec.encode(file)}"

    suspend fun mcp(): List<McpServer>? = Http.get("/claude/mcp")?.jsonArray?.mapNotNull { el ->
        val o = el.jsonObject
        McpServer(
            name = o["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
            type = o["type"]?.jsonPrimitive?.contentOrNull,
            detail = o["detail"]?.jsonPrimitive?.contentOrNull,
            enabled = o["enabled"]?.jsonPrimitive?.booleanOrNull ?: true,
        )
    }

    suspend fun mcpToggle(name: String, enabled: Boolean): ActionResult = actionResult(
        Http.post("/claude/mcp/toggle", buildJsonObject { put("name", name); put("enabled", enabled) })
    )

    private fun actionResult(data: JsonElement?): ActionResult {
        val o = data?.jsonObject ?: return ActionResult(false, "")
        return ActionResult(
            ok = o["ok"]?.jsonPrimitive?.booleanOrNull ?: false,
            message = o["message"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        )
    }

    suspend fun pluginAction(action: String, plugin: String): ActionResult =
        actionResult(Http.post("/claude/plugins/action", buildJsonObject {
            put("action", action)
            put("plugin", plugin)
        }))

    suspend fun marketplaceAction(action: String, target: String): ActionResult =
        actionResult(Http.post("/claude/marketplaces/action", buildJsonObject {
            put("action", action)
            put("target", target)
        }))

    suspend fun catalog(marketplace: String): List<CatalogPlugin>? =
        Http.get("/claude/marketplaces/${UrlCodec.encode(marketplace)}/catalog")?.jsonArray?.mapNotNull { el ->
            val o = el.jsonObject
            CatalogPlugin(
                name = o["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
                description = o["description"]?.jsonPrimitive?.contentOrNull,
                version = o["version"]?.jsonPrimitive?.contentOrNull,
                installed = o["installed"]?.jsonPrimitive?.booleanOrNull ?: false,
            )
        }

    suspend fun mcpAdd(name: String, target: String, transport: String): ActionResult =
        actionResult(Http.post("/claude/mcp", buildJsonObject {
            put("name", name)
            put("target", target)
            put("transport", transport)
        }))

    suspend fun mcpRemove(name: String): ActionResult =
        actionResult(Http.delete("/claude/mcp/${UrlCodec.encode(name)}"))

    suspend fun memories(project: String?): Memories? {
        val query = if (project.isNullOrEmpty()) emptyMap() else mapOf("project" to project)
        val data = Http.get("/claude/memories", query)?.jsonObject ?: return null
        fun parse(key: String) = data[key]?.jsonArray?.mapNotNull { el ->
            val o = el.jsonObject
            Memory(
                scope = o["scope"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
                name = o["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null,
                description = o["description"]?.jsonPrimitive?.contentOrNull,
            )
        } ?: emptyList()
        return Memories(global = parse("global"), project = parse("project"))
    }

    fun memoryUrl(scope: String, project: String?, name: String): String =
        "${Backend.baseUrl}/claude/memories/file?scope=${UrlCodec.encode(scope)}" +
            "&name=${UrlCodec.encode(name)}&project=${UrlCodec.encode(project.orEmpty())}"

    suspend fun deleteMemory(scope: String, project: String?, name: String): Boolean =
        Http.delete("/claude/memories", mapOf(
            "scope" to scope,
            "name" to name,
            "project" to project.orEmpty(),
        )) != null
}

File: tauri/src/lib/services/claudeApi.ts
import { backend, baseUrlOf, type Profile } from "./backend.svelte";
import { http, type HttpClient } from "./http";

export interface Plugin {
  name: string;
  marketplace: string | null;
  version: string | null;
  scope: string | null;
  enabled: boolean;
  description: string | null;
}

export interface Marketplace {
  name: string;
  repo: string | null;
}

export interface Extensions {
  plugins: Plugin[];
  marketplaces: Marketplace[];
}

export interface Skill {
  name: string;
  id: string;
  description: string | null;
  plugin: string | null;
  pluginName: string | null;
  enabled: boolean;
}

export interface McpServer {
  name: string;
  type: string | null;
  detail: string | null;
  enabled: boolean;
}

export interface CatalogPlugin {
  name: string;
  description: string | null;
  version: string | null;
  installed: boolean;
}

export interface Memory {
  scope: string;
  name: string;
  description: string | null;
}

export interface Memories {
  global: Memory[];
  project: Memory[];
}

export interface ActionResult {
  ok: boolean;
  message: string;
}

export interface UsageWindow {
  id: string;
  percent: number;
  resetsAt: string | null;
}

export interface Usage {
  plan: string | null;
  windows: UsageWindow[];
  error: string | null;
}

export interface ServiceComponent {
  name: string;
  status: string;
}

export interface ServiceIncident {
  name: string;
  impact: string;
  status: string;
  latest: string | null;
  updatedAt: string | null;
  shortlink: string | null;
}

export interface ServiceStatus {
  indicator: string;
  description: string;
  components: ServiceComponent[];
  incidents: ServiceIncident[];
  error: string | null;
}

type Wire = Record<string, any>;

const toAction = (data: Wire | null): ActionResult => ({
  ok: data?.ok === true,
  message: data?.message ?? "",
});

export const createClaudeApi = (client: HttpClient, profile: () => Profile) => ({
  async usage(account: string | null = null): Promise<Usage | null> {
    const data = await client.get<Wire>("/claude/usage", account ? { account } : undefined);
    if (!data) return null;
    return {
      plan: data.plan ?? null,
      windows: (data.windows ?? []).map((window: Wire) => ({
        id: window.id ?? "",
        percent: window.percent ?? 0,
        resetsAt: window.resets_at ?? null,
      })),
      error: data.error ?? null,
    };
  },

  async status(): Promise<ServiceStatus | null> {
    const data = await client.get<Wire>("/claude/status");
    if (!data) return null;
    return {
      indicator: data.indicator ?? "none",
      description: data.description ?? "",
      components: (data.components ?? [])
        .filter((item: Wire) => item.name)
        .map((item: Wire) => ({ name: item.name, status: item.status ?? "operational" })),
      incidents: (data.incidents ?? [])
        .filter((item: Wire) => item.name)
        .map((item: Wire) => ({
          name: item.name,
          impact: item.impact ?? "none",
          status: item.status ?? "",
          latest: item.latest ?? null,
          updatedAt: item.updated_at ?? null,
          shortlink: item.shortlink ?? null,
        })),
      error: data.error ?? null,
    };
  },

  async userPrompt(): Promise<string | null> {
    const data = await client.get<Wire>("/claude/prompt");
    return data?.text ?? null;
  },

  async setUserPrompt(text: string): Promise<boolean> {
    return (await client.put("/claude/prompt", { text })) !== null;
  },

  async projectPrompt(project: string): Promise<string | null> {
    const data = await client.get<Wire>("/claude/project-prompt", { project });
    return data?.text ?? null;
  },

  async setProjectPrompt(project: string, text: string): Promise<boolean> {
    return (await client.put("/claude/project-prompt", { project, text })) !== null;
  },

  async extensions(): Promise<Extensions | null> {
    const data = await client.get<Wire>("/claude/plugins");
    if (!data) return null;
    return {
      plugins: (data.plugins ?? [])
        .filter((item: Wire) => item.name)
        .map((item: Wire) => ({
          name: item.name,
          marketplace: item.marketplace ?? null,
          version: item.version ?? null,
          scope: item.scope ?? null,
          enabled: item.enabled !== false,
          description: item.description ?? null,
        })),
      marketplaces: (data.marketplaces ?? [])
        .filter((item: Wire) => item.name)
        .map((item: Wire) => ({ name: item.name, repo: item.repo ?? null })),
    };
  },

  async skills(): Promise<Skill[] | null> {
    const data = await client.get<Wire[]>("/claude/skills");
    if (!Array.isArray(data)) return null;
    return data
      .filter((item) => item.name && item.id)
      .map((item) => ({
        name: item.name,
        id: item.id,
        description: item.description ?? null,
        plugin: item.plugin ?? null,
        pluginName: item.plugin_name ?? null,
        enabled: item.enabled !== false,
      }));
  },

  async skillFiles(plugin: string | null, skillId: string): Promise<string[]> {
    const data = await client.get<Wire>("/claude/skills/files", {
      skill: skillId,
      ...(plugin ? { plugin } : {}),
    });
    return data?.files ?? [];
  },

  skillFileUrl(plugin: string | null, skillId: string, file = "SKILL.md"): string {
    const base = baseUrlOf(profile());
    const pluginPart = plugin ? `&plugin=${encodeURIComponent(plugin)}` : "";
    return `${base}/claude/skills/file?skill=${encodeURIComponent(skillId)}${pluginPart}&file=${encodeURIComponent(file)}`;
  },

  async mcp(): Promise<McpServer[] | null> {
    const data = await client.get<Wire[]>("/claude/mcp");
    if (!Array.isArray(data)) return null;
    return data
      .filter((item) => item.name)
      .map((item) => ({
        name: item.name,
        type: item.type ?? null,
        detail: item.detail ?? null,
        enabled: item.enabled !== false,
      }));
  },

  async mcpToggle(name: string, enabled: boolean): Promise<ActionResult> {
    return toAction(await client.post<Wire>("/claude/mcp/toggle", { name, enabled }));
  },

  async mcpAdd(name: string, target: string, transport: string): Promise<ActionResult> {
    return toAction(await client.post<Wire>("/claude/mcp", { name, target, transport }));
  },

  async mcpRemove(name: string): Promise<ActionResult> {
    return toAction(await client.delete<Wire>(`/claude/mcp/${encodeURIComponent(name)}`));
  },

  async pluginAction(action: string, plugin: string): Promise<ActionResult> {
    return toAction(await client.post<Wire>("/claude/plugins/action", { action, plugin }));
  },

  async marketplaceAction(action: string, target: string): Promise<ActionResult> {
    return toAction(await client.post<Wire>("/claude/marketplaces/action", { action, target }));
  },

  async catalog(marketplace: string): Promise<CatalogPlugin[] | null> {
    const data = await client.get<Wire[]>(`/claude/marketplaces/${encodeURIComponent(marketplace)}/catalog`);
    if (!Array.isArray(data)) return null;
    return data
      .filter((item) => item.name)
      .map((item) => ({
        name: item.name,
        description: item.description ?? null,
        version: item.version ?? null,
        installed: item.installed === true,
      }));
  },

  async memories(project: string | null): Promise<Memories | null> {
    const data = await client.get<Wire>("/claude/memories", project ? { project } : undefined);
    if (!data) return null;
    const parse = (key: string): Memory[] =>
      (data[key] ?? [])
        .filter((item: Wire) => item.scope && item.name)
        .map((item: Wire) => ({ scope: item.scope, name: item.name, description: item.description ?? null }));
    return { global: parse("global"), project: parse("project") };
  },

  memoryUrl(scope: string, project: string | null, name: string): string {
    const base = baseUrlOf(profile());
    return `${base}/claude/memories/file?scope=${encodeURIComponent(scope)}&name=${encodeURIComponent(name)}&project=${encodeURIComponent(project ?? "")}`;
  },

  async deleteMemory(scope: string, project: string | null, name: string): Promise<boolean> {
    return (
      (await client.delete("/claude/memories", { scope, name, project: project ?? "" })) !== null
    );
  },
});

export const claudeApi = createClaudeApi(http, () => backend.active);
